File: src/od/2024e100/rearrange-chars/solution.py
#!/usr/bin/env python3
# Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
# Use of this source is governed by General Public License that can be found
# in the LICENSE file.

def solution1():
    # 创建计数数组
    # 读取输入, 并统计所有字母出现的次数
    # 统计出现的最多次数
    # 然后逆向遍历所有的次数, 并打印结果
    chars = [0] * 256
    for char in input().strip():
        index = ord(char)
        chars[index] += 1

    max_count = max(chars)
    out = []
    for count in range(max_count, 0, -1):
        # 遍历所有的小写字母
        for char_index in range(ord('a'), ord('z') + 1):
            if chars[char_index] == count:
                out.append("{}:{}".format(chr(char_index), count))
        
        # 再遍历所有的大写字母
        for char_index in range(ord('A'), ord('Z') + 1):
            if chars[char_index] == count:
                out.append("{}:{}".format(chr(char_index), count))

    # 打印结果
    s = ";".join(out)
    print(s)

def solution2():
    # 读取输入
    # 创建计数字典
    # 统计所有字母出现的次数
    # 然后转换成数组, 并对其进行排序

    chars = dict()
    for char in input().strip():
        if char not in chars:
            chars[char] = 0
        chars[char] += 1
    print(chars_list)

    # 打印结果
    s = ";".join(F"{char}:{count}" for (char, count) in chars_list)
    print(s)

def main():
    solution1()

if __name__ == "__main__":
    main()

File: src/od/2024e100/rearrange-chars/solution.cpp
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

#include <algorithm>
#include <array>
#include <iostream>
#include <vector>

int main() {
  // 创建计数数组
  // 读取输入, 并统计所有字母出现的次数
  // 统计出现的最多次数
  // 然后逆向遍历所有的次数, 并打印结果

  const size_t kNumChars = 256;
  std::array<int, kNumChars> chars;
  for (int i = 0; i < kNumChars; ++i) {
    chars[i] = 0;
  }

  std::string line;
  std::getline(std::cin, line);
  for (char chr : line) {
    int index = static_cast<int>(chr);
    chars[index] += 1;
  }

  const int max_count = *std::max(chars.cbegin(), chars.cend());
  std::vector<std::string> out;
  const size_t kBufLen = 64;
  char buf[kBufLen];

  for (int count = max_count; count > 0; --count) {
    // 遍历所有的小写字母
    for (int char_index = static_cast<int>('a'); char_index <= static_cast<int>('z'); ++char_index) {
      if (chars[char_index] == count) {
        const int s_len = std::snprintf(buf, kBufLen, "%c:%d", static_cast<char>(char_index), count);
        out.emplace_back(buf, s_len);
      }
    }

    // 再遍历所有的大写字母
    for (int char_index = static_cast<int>('A'); char_index <= static_cast<int>('Z'); ++char_index) {
      if (chars[char_index] == count) {
        const int s_len = std::snprintf(buf, kBufLen, "%c:%d", static_cast<char>(char_index), count);
        out.emplace_back(buf, s_len);
      }
    }
  }

  // 打印结果
  for (int i = 0; i + 1 < out.size(); ++i) {
    std::cout << out[i] << ";";
  }
  std::cout << out[out.size() - 1] << std::endl;

  return 0;
}
